docs/source/developer/concepts.rst

```diff
@@ -0,0 +1,307 @@
+
+.. code::
+
+    template <class E>
+    derived_type& operator=(const xexpression<E>&);
+
+Like computed assignment operators, it evaluates the expression inside a temporary before calling the ``assign``
+method. Classes inheriting from ``xsemantic_base`` must redeclare this method either in their protected section
+(if they are not final classes) or in their public section. In both cases, they should forward the call to their
+base class.
+
+Two refinements of this concept are provided, ``xcontainer_semantic`` and ``xview_semantic``. Refer to the
+:ref:`xtensor-assign-label` section for more details about semantic classes and how they're involved in expression
+assignment.
+
+xsemantic classes hierarchy:
+
+.. image:: xsemantic_classes.svg
+
+.. _xcontainer-concept-label:
+
+xcontainer
+~~~~~~~~~~
+
+The ``xcontainer`` class provides methods for container-based expressions. It does not hold any data, this is delegated
+to inheriting classes. It assumes the data are stored using a strided-index scheme. ``xcontainer`` defines the following
+methods:
+
+**Shape, strides and size**
+
+.. code::
+
+    size_type size() const noexcept;
+    size_type dimension() const noexcept;
+
+    const inner_shpae_type& shape() const noexcept;
+    const inner_strides_type& strides() const noexcept;
+    const inner_backstrides_type& backstrides() const noexcept;
+
+**Data access methods**
+
+.. code::
+
+    template <class... Args>
+    const_reference operator()(Args... args) const;
+
+    template <class... Args>
+    const_reference at(Args... args) const;
+
+    template <class S>
+    disable_integral_t<S, const_reference> operator[](const S& index) const;
+
+    template <class I>
+    const_reference operator[](std::initialize_list<I> index) const;
+
+    template <class It>
+    const_reference element(It first, It last) const;
+
+    const container_type& data() const;
+
+(and their non-const counterpart)
+
+**Broadcasting methods**
+
+.. code::
+
+    template <class S>
+    bool broadcast_shape(const S& shape) const;
+
+    template <class S>
+    bool is_trivial_broadcast(const S& strides) const;
+
+Lower-level methods are also provided, meant for optimized assignment and BLAS bindings.
+They are covered in the :ref:`xtensor-assign-label` section.
+
+If you read the entire code of ``xcontainer``, you'll notice that two types are defined for shape,
+strides and backstrides: ``shape_type`` and ``inner_shape_type``, ``strides_type`` and
+``inner_strides_type``, and ``backstrides_type`` and ``inner_backstrides_type``.
+
+The inner types are used for storing data, while the other ones are used for reshape and resize
+APIs. This way, it is possible to write in C++ a wrapper of a python structure that holds the shape,
+strides and backstrides allocated on the python side, and provide methods for resizing and reshaping
+that accept structures allocated on the C++ side.
```

Comment:
The distinction between `inner_shape_type` and `shape_type` was motivated by the `xtensor-python` wrapper around numpy data structures, where the inner shape type is a proxy on the shape section of the numpy array object. It cannot have a value semantics on its own as it is bound to the entire numpy array.